ONSdigital/rm-collection-exercise-service
# Refactor sample validation

# Motivation and Context
The 'validation' of the samples was spaghetti code and highly inefficient.

Given the need to load large samples for ASHE etc. it seemed prudent to refactor all the code that the lies on the sample load pipeline, to make it more performant and reliable.

In the course of the refactoring, a number of bugs were spotted, such as inability to recover from momentary connection problems with the Survey, Collection Instrument and Party services - these should retry without manual intervention.

The code was inefficient, badly performant, hard to understand, maintain, and it did not handle errors well.

# What has changed
Major refactoring of the `ValidateSampleUnits` class.

Sadly it's still not quite honouring the single responsibility principle, but the code is a lot less complex and should make future refactoring easier.

Substantial improvement to performance and error handling.

# How to test?
Load a big sample. Make sure the sample validates OK and the collection exercise changes to "Ready for Live" status.

# Links
Trello: https://trello.com/c/eAmhl3iB/729-use-stream-and-multithreading-in-validatesampleunits-to-rapidly-validate-the-samples-13
